## Simulación Monty-Hall
Integrante 1: Juan Enrique Aguirre López

Integrante 2: Pablo García Bedoy Fernández 

# Objetivos

### Objetivo General
* Modelar el problema de Monty-Hall

### Objetivos específicos
* Llegar a una conclusión de ver si, en realidad, es más exitosa la estrategia de cambiar de puerta que seguir con la misma puerta
* Utilizar numeros aleatorios con el objetivo de simular un gran número de iteraciones, para obtener la probabilidad de éxito al cambiar de puerta o al mantener la misma

## Modelo que representa el problema

La situación es la siguiente, se tienen tres puertas y un concursante que puede ganar un premio (un coche). Se le pide al concursante escoger una de las tres puertas, ![](montyhall_3doors.png)

al hacerlo, se descubre una de las puertas donde no hay premio. ![](montyhall_3doors4.png)

En este momento se le pregunta al concursante si desea cambiar la elección de la puerta. La pregunta está en si es más probable conseguir el premio si cambia o no de puerta. 

![](montyhall_3doors2.png)

Cambiar resulta ser la estrategia “más” ganadora, y la razón es la siguiente: Al escoger por primera vez se tiene una probabilidad de  1/3  de escoger la puerta correcta, y hay una probabilidad de  2/3  de no haberlo hecho, es decir, en las otras dos puertas que no fueron escogidas, hay un  2/3  de probabilidad de que en alguna de ellas está, por lo tanto, al descubrir una de estas, la probabilidad de  2/3  se mantiene pero ahora solo en una puerta, la cerrada que no fue escogida. 

![](montyhall_3doors3.png)

Por lo tanto, es dos veces más probable de conseguir el premio si se cambia la elección inicial por la puerta cerrada restante.

## Solución del problema

#### Primero presentaremos como sería la solución del problema realizandose "N" numero de iteraciones

In [6]:
import random

success = 0
switch = random.randint(0,1)

N = 10000
for i in range(N):
    car = random.randint(0,2)
    choice = random.randint(0,2)
    door_to_reveal = random.randint(0,2)

    while door_to_reveal == car or choice == door_to_reveal:
        door_to_reveal = random.randint(0,2)

    final_choice = 0	
    if switch == 1:
        while final_choice == choice or final_choice == door_to_reveal:
            final_choice += 1
    else:
        final_choice = choice

    success += final_choice == car

if switch == 1:
	print("\nYou switched doors...\n")
else:
	print("\nYou did not switch doors...\n")
print("\nYour win rate was: ", float(success)/N, "\n")

print(switch)


You switched doors...


Your win rate was:  0.6593 

1


**Cada vez que se corra este programa nos dará las posibilidades de éxito (de sacarnos el carro) en base a la decisión de cambiar o no de puerta.**

### A continuación, presentaremos estos resultados en una tabla para que sean más fácil de visualizar los resultados

**Empezaremos con definir este programa como una función, una función en la que siempre se cambie de puerta (Monty_hall2(N)), y otra en la que se quede siempre con la misma puerta (Monty_hall(N)).**

In [7]:
import random

def Monty_hall(N):
    switch = 0
    success = 0
    for i in range(N):
        car = random.randint(0,2)
        choice = random.randint(0,2)
        door_to_reveal = random.randint(0,2)

        while door_to_reveal == car or choice == door_to_reveal:
            door_to_reveal = random.randint(0,2)

        final_choice = 0	
        if switch == 1:
            while final_choice == choice or final_choice == door_to_reveal:
                final_choice += 1
        else:
            final_choice = choice

        success += final_choice == car
    return (float(success)/N)

def Monty_hall2(N):
    switch = 1
    success = 0
    for i in range(N):
        car = random.randint(0,2)
        choice = random.randint(0,2)
        door_to_reveal = random.randint(0,2)

        while door_to_reveal == car or choice == door_to_reveal:
            door_to_reveal = random.randint(0,2)

        final_choice = 0	
        if switch == 1:
            while final_choice == choice or final_choice == door_to_reveal:
                final_choice += 1
        else:
            final_choice = choice

        success += final_choice == car
    return (float(success)/N)

**Una vez que ya establecimos las funciones para cuando cambias de puerta o para cuando te quedas con la misma, apreciaremos qué es lo que pasa cuando lo hacemos para pocos casos y qué es lo que sucede con la probabilidad cuando aumentamos el número de casos.**

In [10]:
import numpy as np
import pandas as pd
import sympy as sym
sym.init_printing(use_latex='mathjax')


N=np.logspace(1,5,5).astype(int)

df= pd.DataFrame(index=N)

for n in N:
    df.loc[n,'Probabilidad de éxito sin cambiar'] = Monty_hall(n)
    df.loc[n,'Probabilidad de éxito con cambio'] = Monty_hall2(n)
df

Probabilidad de éxito sin cambiar  Probabilidad de éxito con cambio
10                                0.40000                           0.70000
100                               0.32000                           0.76000
1000                              0.35800                           0.68200
10000                             0.33010                           0.65970
100000                            0.33258                           0.66873

## Conclusiones

En base a el código anterior, se pudo obtener las probababilidades de éxito que habíaal cambiar de puerta y al quedarnos con la misma puerta. Siendo los resultados obtenidos coherentes con los objetivos planteados y la literatura e investigación previa. Los resultados que obtuvimos, a través de realizar las aproximaciones con N número de casos aleatorios, fue que si decidias cambiar de puerta la probabilidad es $\approx$ 66%. Mientras que si no cambias de puerta, es decir, te quedas con la misma la probabilidad es $\approx$ 33%.

## Referencias

Bartoldson, B. (2018). Monty_Hall_Simulation.py. Recuperado de https://github.com/bbartoldson/examples/blob/master/Monty%20Hall%20Simulation/Monty_Hall_Simulation.py


S.A. (2015). Modeling the Monty Hall Problem with Python. Recuperado de http://code.jasonbhill.com/python/modeling-the-monty-hall-problem-with-python/